src/lib/field.unit-test.ts

```diff
@@ -0,0 +1,317 @@
+import { ProvablePure } from '../snarky.js';
+import { Field } from './core.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { test, Random } from './testing/property.js';
+import { deepEqual } from 'node:assert/strict';
+import { Provable } from './provable.js';
+import { Binable } from '../bindings/lib/binable.js';
+import { ProvableExtended } from './circuit_value.js';
+import { throws } from 'node:assert';
+import { FieldType } from './field.js';
+
+// types
+Field satisfies Provable<Field>;
+Field satisfies ProvablePure<Field>;
+Field satisfies ProvableExtended<Field>;
+Field satisfies Binable<Field>;
+
+// constructor
+test(Random.field, Random.json.field, (x, y, assert) => {
+  let z = Field(x);
+  assert(z instanceof Field);
+  assert(z.toBigInt() === x);
+  assert(z.toString() === x.toString());
+  assert(z.isConstant());
+  deepEqual(z.toConstant(), z);
+
+  assert((z = new Field(x)) instanceof Field && z.toBigInt() === x);
+  assert((z = Field(z)) instanceof Field && z.toBigInt() === x);
+  assert((z = Field(z.value)) instanceof Field && z.toBigInt() === x);
+
+  z = Field(y);
+  assert(z instanceof Field);
+  assert(z.toString() === y);
+  deepEqual(Field.fromJSON(y), z);
+  assert(z.toJSON() === y);
+});
+
+// special generator
+let SmallField = Random.reject(
+  Random.field,
+  (x) => x.toString(2).length > Fp.sizeInBits - 2
+);
+
+// arithmetic, both in- and outside provable code
+equivalent2((x, y) => x.add(y), Fp.add);
```

Comment:
the idea of these tests is that you have two implementations - the TS finite field `Fp`, and `Field` - and you test that they are equivalent, both in- and outside provable code. "Equivalent" also means that one throws an error if and only if the other throws an error